lib/iris/_concatenate.py

```diff
@@ -1240,35 +1262,33 @@ def _build_aux_factories(self, dim_coords_and_dims, aux_coords_and_dims):
         old_aux_coords = [a[0] for a in cube_signature.aux_coords_and_dims]
         new_dim_coords = [d[0] for d in dim_coords_and_dims]
         new_aux_coords = [a[0] for a in aux_coords_and_dims]
-        scalar_coords = cube_signature.scalar_coords
+        old_scalar_coords = cube_signature.scalar_coords
+        new_scalar_coords = scalar_coords
 
         aux_factories = []
 
         # Generate all the factories for the new concatenated cube.
         for i, (coord, dims, factory) in enumerate(
             cube_signature.derived_coords_and_dims
         ):
-            # Check whether the derived coordinate of the factory spans the
-            # nominated dimension of concatenation.
-            if self.axis in dims:
```

Comment:
I actually think that it **is** necessary to remove this `if` statement.

To be honest, when first coding it, I just copy-pasted this from the other `_build_...` methods (e.g., [here](https://github.com/SciTools/iris/blob/13b2a7466c71121bf5d3261f3ddf9803b46ca694/lib/iris/_concatenate.py#L1094)) without too much thought. In these cases, coordinates that span the concatenation axis need [special treatment](https://github.com/SciTools/iris/blob/13b2a7466c71121bf5d3261f3ddf9803b46ca694/lib/iris/_concatenate.py#L1096-L1131); others that do not span it are simply [copied](https://github.com/SciTools/iris/blob/13b2a7466c71121bf5d3261f3ddf9803b46ca694/lib/iris/_concatenate.py#L1133) from the original cube.

However, the case of building aux factories is different: **all** coordinates of the final concatenated cube are already built (and passed to the function as `dim_coords_and_dims`, `aux_coords_and_dims`, and `scalar_coords`). We now have to make sure that **all** aux factories (regardless if they span the concatenation axis or not) are properly updated with these final coordinates, so we don't need a special treatment. Otherwise the dependencies could still contain references to old coordinates, which then causes errors like the one described in #5339.

Reply:
>To be honest, when first coding it, I just copy-pasted this from the other `_build_...` methods

To be honest, when _reviewing_ it, I also 'copy-pasted' my thoughts by basing on the other `_build_...` methods!

Thanks for the explanation. Aux factories are a tricky topic. I think others in my situation - reading unfamiliar code - could fall into similar trap when they compare `_build_aux_factories()` with the other adjacent methods. Do you think you could make clear in a comment why it's important to update **all** aux factories, regardless of spanning?

Reply:
> Do you think you could make clear in a comment why it's important to update all aux factories, regardless of spanning?

I tried to do that in https://github.com/SciTools/iris/pull/5340/commits/6da8979fc7a0e2dad3f5ca101a1debaa307d1ca7, please let me know what you think!

Reply:
A little long, but very informative! I'm happy to go with it.